File: apple/AgentDeck/Model/AgentState.swift
// AgentState.swift — State enums & prompt types
// Ported from shared/src/states.ts

import Foundation

// MARK: - State

enum AgentConnectionState: String, Codable, Sendable, CaseIterable {
    case disconnected
    case idle
    case processing
    case awaitingPermission = "awaiting_permission"
    case awaitingOption = "awaiting_option"
    case awaitingDiff = "awaiting_diff"

    var isAwaiting: Bool {
        switch self {
        case .awaitingPermission, .awaitingOption, .awaitingDiff: true
        default: false
        }
    }

    var isActive: Bool {
        self == .processing
    }

    var displayLabel: String {
        switch self {
        case .disconnected: "DISCONNECTED"
        case .idle: "IDLE"
        case .processing: "PROCESSING"
        case .awaitingPermission: "PERMISSION"
        case .awaitingOption: "SELECT"
        case .awaitingDiff: "DIFF REVIEW"
        }
    }
}

// MARK: - Permission Mode

enum PermissionMode: String, Codable, Sendable {
    case `default`
    case plan
    case acceptEdits
    case dontAsk
    case bypassPermissions
}

// MARK: - Prompt Option

struct PromptOption: Codable, Sendable, Identifiable {
    let index: Int
    let label: String
    var shortcut: String?
    var recommended: Bool?
    var selected: Bool?
    var kind: String? = nil

    var id: Int { index }

    var isFreeformInput: Bool { kind == "freeform_input" }
}

// MARK: - Prompt Type

enum PromptType: String, Codable, Sendable {
    case yesNo = "yes_no"
    case yesNoAlways = "yes_no_always"
    case multiSelect = "multi_select"
    case diffReview = "diff_review"
}

// MARK: - Dashboard State (composite observable state)

struct DashboardState: Sendable {
    // Connection
    var bridgeConnected = false
    /// Session that produced the latest state/update payload. This is used
    /// for attribution and may change automatically as hooks arrive.
    var sessionId: String?
    /// Session explicitly focused by the user via a session row, creature tap,
    /// menubar row, or hardware session switch. Visual selection should use
    /// this field, not `sessionId`, so activity does not look like selection.
    var focusedSessionId: String?

    // Agent state
    var state: AgentConnectionState = .disconnected
    var permissionMode: PermissionMode = .default
    var agentType: String?  // "claude-code" | "openclaw"
    var agentCapabilities: AgentCapabilities?

    // Tool info
    var currentTool: String?
    var toolInput: String?
    var toolProgress: String?

    // Project / Model
    var projectName: String?
    var modelName: String?
    var effortLevel: String?
    var billingType: BillingType = .unknown

    // Prompt
    var options: [PromptOption] = []
    var promptType: PromptType?
    var question: String?
    var navigable = false
    var cursorIndex = 0
    var suggestedPrompt: String?

    // Model catalog
    var modelCatalog: [ModelCatalogEntry] = []
    var sessionStatus: OcSessionStatus?

    // Remote / Gateway
    var remoteUrl: String?
    var pairingUrl: String?
    var workerSessionCount: Int?
    var gatewayAvailable = false
    var gatewayConnected = false
    var gatewayHasError = false
    var gatewayAuthStatus: String?
    var gatewayAuthRequestId: String?
    var gatewayAuthMessage: String?
    /// Locally-generated Gateway identity (Ed25519 public-key SHA-256 hex).
    /// Surfaced in pairing hints when auth is blocked on signature failure.
    var gatewayDeviceId: String?
    var daemonPort: Int?

    // Usage
    var sessionDurationSec = 0
    var inputTokens = 0
    var outputTokens = 0
    var toolCalls = 0
    var estimatedCostUsd: Double?
    var sessionPercent: Double?
    var costSpent: Double?
    var costLimit: Double?
    var resetTime: String?
    var resetDate: String?
    var fiveHourPercent: Double?
    var fiveHourResetsAt: String?
    var sevenDayPercent: Double?
    var sevenDayResetsAt: String?
    /// Per-model scoped weekly caps (e.g. the "Fable" cap) — distinct from the
    /// account-wide 5h/7d. Consumed via the daemon relay only (relay-only boundary).
    var scopedLimits: [ScopedUsageLimit]?
    var previousFiveHourPercent: Double?
    var previousSevenDayPercent: Double?
    var extraUsageEnabled: Bool?
    var extraUsageMonthlyLimit: Double?
    var extraUsageUsedCredits: Double?
    var extraUsageUtilization: Double?
    var oauthConnected: Bool?
    var ollamaStatus: OllamaStatus?
    var usageStale: Bool?
    var codexAuthMode: String?
    var codexWebAuthConnected: Bool?
    var codexPlanType: String?
    var codexAccountId: String?
    var codexSubscriptionActiveUntil: String?
    var codexLastRefreshAt: String?
    var codexRateLimits: CodexRateLimits?
    var mlxModels: [String] = []
    var mlxModelCatalog: [String] = []
    var subscriptions: [SubscriptionInfo] = []
    var antigravityStatus: AntigravityStatusInfo?

    // Anthropic Admin API (optional — org-wide token usage when user
    // pastes a Console Admin API key in Settings). Separate from
    // `fiveHourPercent` / `sevenDayPercent` above (Pro/Max subscription
    // quota) because API usage billing is a different metric tier.
    var adminApiKeyPresent: Bool = false
    var adminApiTodayInputTokens: Int?
    var adminApiTodayOutputTokens: Int?
    var adminApiTodayCacheReadTokens: Int?
    var adminApiTodayCacheCreationTokens: Int?
    var adminApiMonthInputTokens: Int?
    var adminApiMonthOutputTokens: Int?
    var adminApiMonthCacheReadTokens: Int?
    var adminApiMonthCacheCreationTokens: Int?
    var adminApiTopModels: [AdminApiModelUsage] = []
    var adminApiFetchedAt: Double?
    var adminApiStale: Bool?

    // Voice
    var voiceState: String?  // idle | recording | transcribing | error
    var voiceText: String?
    var voiceError: String?

    // Voice Assistant (wake word pipeline)
    var voiceAssistantState: String?  // idle | listening | processing | speaking | disabled
    var voiceAssistantText: String?
    var voiceAssistantResponseText: String?

    // Display
    var hostDisplayOn = true

    // Multi-session
    var siblingSessions: [SessionInfo] = []
    /// Whether a `sessions_list` has ever arrived. `siblingSessions` alone
    /// cannot distinguish "no list yet" from "list arrived and is empty", and
    /// the Pixoo `_primary` fallback needs that difference — see
    /// PixooRenderer.syncCreatures. Defaults to false ("not received yet") so
    /// any surface that never sets it keeps the pre-list fallback behavior.
    var sessionsListReceived = false

    // Device module health (from daemon statusSnapshot aggregation)
    var moduleHealth: ModuleHealthState?

}

// MARK: - Anthropic Admin API

struct AdminApiModelUsage: Sendable, Codable, Equatable {
    let model: String
    let totalTokens: Int
}

// MARK: - Module Health

struct ModuleHealthState: Sendable {
    var adb: AdbHealth?
    var d200h: D200hHealth?
    var pixoo: PixooHealth?
    var serial: SerialHealth?
    var streamDeck: StreamDeckHealth?
    /// Divoom Timebox Mini (11×11 BLE) — daemon `statusSnapshot()`.
    var timebox: BLEMatrixHealth?
    /// iDotMatrix (32×32 BLE) — daemon `statusSnapshot()`.
    var idotmatrix: BLEMatrixHealth?
    /// Wi-Fi WebSocket e-ink panels (XTeink X3 …) that registered via
    /// `client_register {clientType:"eink-device"}`. Same volunteer-roster model
    /// as Stream Deck: present only while the panel's WS is live.
    var eink: EinkHealth?
    /// Android dashboard apps (tablet / e-ink launcher) that registered via
    /// `client_register {clientType:"android-dashboard"}` over Wi-Fi WS.
    /// Same volunteer-roster model; covers devices ADB never sees (Swift
    /// daemon has no ADB, and WiFi-only tablets have no USB bridge).
    var androidDashboards: AndroidDashboardHealth?
    /// TUI dashboards (`agentdeck dashboard`) that registered via
    /// `client_register {clientType:"tui"}`. Live-presence roster — a row
    /// exists exactly while the terminal client's WS is open.
    var tuiDashboards: TuiDashboardHealth?
    /// WiFi-WS ESP32 boards (announced `device_info` over their WebSocket).
    /// Both daemons emit `esp32Wifi` with per-board `serialActive` so the rail
    /// can suppress boards already shown as USB-serial rows (single-path
    /// transport dedup — serial drives, WiFi is a hot standby).
    var esp32Wifi: Esp32WifiHealth?
}

/// Shared shape for the BLE matrix panels (Timebox Mini, iDotMatrix). Both
/// daemon modules (`TimeboxModule`, `IDotMatrixModule`) emit an identical
/// `statusSnapshot()` — connection state plus a human `statusReason` ("connected",
/// "connecting…", "retrying (backed off)", "paused: host display asleep", or an
/// error string) — so the topology rail can show *why* a panel isn't streaming
/// instead of silently omitting it.
struct BLEMatrixHealth: Sendable {
    var configuredDeviceCount: Int = 0
    var connected: Bool = false
    var deviceName: String?
    var statusReason: String?
    var displayDimmed: Bool = false
    var hasFrame: Bool = false
    var lastError: String?
}

struct StreamDeckHealth: Sendable {
    /// Physical Stream Deck devices the Elgato plugin is driving, as
    /// reported via the `client_register` announcement.
    var devices: [StreamDeckDeviceInfo] = []
}

/// Wi-Fi WebSocket e-ink panels (XTeink X3 …) that registered as
/// `clientType:"eink-device"`. Same volunteer-roster model as Stream Deck.
struct EinkHealth: Sendable {
    var devices: [EinkDeviceInfo] = []
}

struct EinkDeviceInfo: Sendable, Hashable {
    var id: String
    var name: String
    /// Panel family ("eink"). Kept as a String for forward compatibility.
    var family: String?
    /// Native panel resolution as reported by the firmware (columns×rows pixels).
    var columns: Int?
    var rows: Int?
}

/// Android dashboard apps connected over Wi-Fi WS that registered as
/// `clientType:"android-dashboard"`. Same volunteer-roster model as Stream Deck.
struct AndroidDashboardHealth: Sendable {
    var devices: [AndroidDashboardDeviceInfo] = []
}

struct AndroidDashboardDeviceInfo: Sendable, Hashable {
    var id: String
    var name: String
    /// "tablet" | "eink" — how the app classified its own hardware
    /// (EinkDetector). Kept as a String for forward compatibility.
    var kind: String?
}

/// TUI dashboards (`agentdeck dashboard`) connected over WS that registered as
/// `clientType:"tui"`. Same volunteer-roster model as Stream Deck.
struct TuiDashboardHealth: Sendable {
    var devices: [TuiClientInfo] = []
}

struct TuiClientInfo: Sendable, Hashable {
    /// Stable per-process id (`hostname#pid`) so two TUIs on one host render
    /// as two rows.
    var id: String
    /// Host name the TUI runs on.
    var name: String
}

/// WiFi-WS ESP32 boards. `serialActive == true` means the same physical board
/// is currently driven over USB serial (its WiFi socket is a hot standby) —
/// the rail suppresses those to avoid double rows for one device.
struct Esp32WifiHealth: Sendable {
    var devices: [WifiEsp32DeviceInfo] = []
}

struct WifiEsp32DeviceInfo: Sendable, Hashable {
    var board: String
    var ip: String?
    var version: String?
    /// Node daemon ages entries out (90 s without device_info → stale);
    /// the Swift daemon evicts on socket close so this stays false there.
    var stale: Bool = false
    var serialActive: Bool = false
}

struct StreamDeckDeviceInfo: Sendable, Hashable {
    var id: String
    var name: String
    /// Canonical family identifier derived from Elgato's DeviceType enum. Kept
    /// as a String so future Elgato families decode without a model update.
    var family: String?
    var columns: Int?
    var rows: Int?
}

struct AdbHealth: Sendable {
    var available: Bool = false
    var devices: [String] = []
    var classifiedDevices: [ClassifiedDevice] = []
    var reverseReadyCount: Int = 0
    var lastError: String?
}

struct ClassifiedDevice: Sendable, Hashable {
    var serial: String
    var manufacturer: String?
    var model: String?
    /// Raw class string from the wire (e.g. `"e-ink.crema"`, `"ulanzi.tc001"`,
    /// `"android.tablet"`). Kept as String rather than enum so older
    /// daemons that emit unknown class names still decode.
    var deviceClass: String
}

/// What kind of Android device is attached. Used for raw-value constants
/// in the UI; wire format is a plain string so unknown classes from
/// newer daemons still round-trip.
///
/// - `.eInkCrema` / `.eInkPantone` / `.eInkKobo`: e-ink devices that need
///   slow, low-contrast UI and avoid animation. Wrong refresh strategy =
///   ghosting + flicker.
/// - `.androidTablet`: everything else (Lenovo, dev phones, generic).
///   Full-colour UI, normal refresh.
///
/// The Ulanzi TC001 8×32 LED matrix is intentionally absent: it is an ESP32
/// board (env `led8x32`) driven over USB serial / WiFi WS and surfaces through
/// the serial pipeline (`esp32DisplayName`), never the ADB tier. The legacy
/// `.ulanziTc001` case was removed on 2026-06-25.
///
/// Not gated by `#if os(macOS)` because the iOS companion's topology view
/// consumes the same wire format.
enum AdbDeviceClass: String, Sendable {
    case eInkCrema = "e-ink.crema"
    case eInkPantone = "e-ink.pantone"
    case eInkKobo = "e-ink.kobo"
    case androidTablet = "android.tablet"
}

struct D200hHealth: Sendable {
    var connected: Bool = false
}

struct PixooHealth: Sendable {
    var configuredDeviceCount: Int = 0
    var deviceIps: [String] = []
    var hasFrame: Bool = false
    var displayDimmed: Bool = false
    var lastPushError: String?
    var devices: [PixooDeviceHealth] = []
}

struct PixooDeviceHealth: Sendable {
    var ip: String
    var online: Bool
    var failures: Int
    var backedOff: Bool
}

struct SerialHealth: Sendable {
    var connectedPorts: [String] = []
    var connectedBoards: [SerialPortInfo] = []
    var lastError: String?
}

struct SerialPortInfo: Sendable, Hashable {
    var port: String
    var board: String?
    var firmwareVersion: String?
    /// Board's own WiFi STA state from `device_info` — surfaced on the serial
    /// row so a dual-homed board reads "USB · WiFi" instead of growing a
    /// second row (see `Esp32WifiHealth.serialActive`).
    var wifiConnected: Bool?
}

extension SerialPortInfo {
    /// Human-friendly label for an ESP32 `board` wire string
    /// (`DeviceInfoMessage.board`). The descriptive suffix mirrors the friendly
    /// column in docs/hardware-compatibility.md (the naming SSOT). Unknown or
    /// missing boards fall back to a plain "ESP32" during the ~2s window before
    /// the firmware's first `device_info` frame arrives.
    ///
    /// This is the single source of truth — `TopologyRail` and
    /// `MenuBarTopologyList` both call it so the two surfaces can't drift.
    static func esp32DisplayName(for board: String?) -> String {
        switch board {
        case "ips_35":         return "ESP32 · IPS 3.5\""
        case "ips_10":         return "ESP32 · IPS 10.1\""
        case "round_amoled":   return "ESP32 · Round AMOLED 1.8\""
        case "86box":          return "ESP32 · 86 Box 4\""
        case "ttgo_t_display": return "ESP32 · TTGO T-Display 1.14\""
        // XTeink X3/X4 are ESP32-C3 e-ink readers running the external
        // CrossPoint fork; surface the product name, not the raw board string.
        case "xteink_x3":      return "XTeink X3"
        case "xteink_x4":      return "XTeink X4"
        // Ulanzi TC001 is an ESP32 under the hood but sold as a finished
        // product, so surface the brand instead of the raw board name.
        case "ulanzi_tc001":   return "Ulanzi TC001"
        case .some(let b) where !b.isEmpty: return "ESP32 · \(b)"
        default:               return "ESP32"
        }
    }
}

File: android/app/src/main/kotlin/dev/agentdeck/state/AgentState.kt
package dev.agentdeck.state

import dev.agentdeck.net.AgentCapabilities
import dev.agentdeck.net.AgentState
import dev.agentdeck.net.BridgeConnection
import dev.agentdeck.net.BridgeEvent
import dev.agentdeck.net.DimConfig
import dev.agentdeck.net.ModelCatalogEntry
import dev.agentdeck.net.ModuleHealthState
import dev.agentdeck.net.OcSessionStatus
import dev.agentdeck.net.OllamaStatus
import dev.agentdeck.net.PermissionMode
import dev.agentdeck.net.PromptOption
import dev.agentdeck.net.SessionInfo
import dev.agentdeck.net.SubscriptionInfo
import dev.agentdeck.net.AntigravityStatusInfo
import dev.agentdeck.net.CodexRateLimits
import dev.agentdeck.net.StateUpdate
import dev.agentdeck.net.UsageUpdate
import dev.agentdeck.net.VoiceState
import dev.agentdeck.net.toTimelineEntry
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

private const val TAG = "Terrarium"
private const val VERBOSE_STATE_LOGS = false
private val AGGREGATE_AGENT_TYPES = setOf("daemon", "openclaw")

private inline fun stateDebug(message: () -> String) {
    if (VERBOSE_STATE_LOGS || Log.isLoggable(TAG, Log.DEBUG)) {
        Log.d(TAG, message())
    }
}

data class DashboardState(
    val agentState: AgentState = AgentState.DISCONNECTED,
    val permissionMode: PermissionMode = PermissionMode.DEFAULT,
    val agentType: String? = null,
    val currentTool: String? = null,
    val toolInput: String? = null,
    val toolProgress: String? = null,
    val projectName: String? = null,
    val modelName: String? = null,
    val effortLevel: String? = null,
    val billingType: String? = null,
    val options: List<PromptOption> = emptyList(),
    val promptType: String? = null,
    val question: String? = null,
    val suggestedPrompt: String? = null,
    val remoteUrl: String? = null,
    val usage: UsageUpdate = UsageUpdate(),
    val voice: VoiceState = VoiceState(),
    val sessionId: String? = null,
    val bridgeConnected: Boolean = false,
    val agentCapabilities: AgentCapabilities? = null,
    val modelCatalog: List<ModelCatalogEntry>? = null,
    val sessionStatus: OcSessionStatus? = null,
    val pairingUrl: String? = null,
    val navigable: Boolean? = null,
    val cursorIndex: Int? = null,
    val hostDisplayOn: Boolean = true,
    /** Host's dim instruction (off/min/level). null ⇒ legacy full-off. */
    val hostDim: DimConfig? = null,
    /**
     * Session explicitly focused by the user, broadcast by the daemon's
     * focus relay. Use this — not [sessionId] — for visual selection
     * (the daemon connection's [sessionId] is a bridge-core UUID, not a
     * real session id).
     */
    val focusedSessionId: String? = null,
    val siblingSessions: List<SessionInfo> = emptyList(),
    val workerSessionCount: Int? = null,
    val oauthConnected: Boolean? = null,
    val ollamaStatus: OllamaStatus? = null,
    val gatewayAvailable: Boolean? = null,
    val gatewayConnected: Boolean? = null,
    val gatewayHasError: Boolean? = null,
    val moduleHealth: ModuleHealthState? = null,
    val voiceAssistantState: String? = null,
    val voiceAssistantText: String? = null,
    val voiceAssistantResponseText: String? = null,
    val mlxModels: List<String> = emptyList(),
    val subscriptions: List<SubscriptionInfo> = emptyList(),
    val antigravityStatus: AntigravityStatusInfo? = null,
    /**
     * Codex (ChatGPT) usage limits — hoisted from usage_update so a later
     * usage event carrying null doesn't wipe the last known value. Only
     * rides on usage_update (never state_update), exactly like macOS/iOS.
     */
    val codexRateLimits: CodexRateLimits? = null,
)

class AgentStateHolder private constructor() {

    companion object {
        val instance: AgentStateHolder by lazy { AgentStateHolder() }
    }

    private val _state = MutableStateFlow(DashboardState())
    val state: StateFlow<DashboardState> = _state.asStateFlow()

    /** Last known state for offline cache display */
    private var lastKnownState: DashboardState? = null

    init {
        BridgeConnection.instance.onEvent = ::handleEvent
    }

    fun getLastKnownState(): DashboardState? = lastKnownState

    private fun handleEvent(event: BridgeEvent) {
        when (event) {
            is BridgeEvent.State -> {
                stateDebug { "StateEvent: state=${event.data.state}, agentType=${event.data.agentType}, tool=${event.data.currentTool}" }
                _state.update { current ->
                    // Keep aggregate identity when current is daemon/openclaw and incoming event is from a coding agent.
                    // No need to wait for siblingSessions to be populated — the first state_update may arrive
                    // before sessions_list, and we must preserve the aggregate sessionId from the start.
                    val keepAggregateIdentity =
                        current.agentType in AGGREGATE_AGENT_TYPES &&
                            event.data.agentType != null &&
                            event.data.agentType !in AGGREGATE_AGENT_TYPES
                    val resolvedAgentType = when {
                        keepAggregateIdentity -> current.agentType
                        else -> event.data.agentType ?: current.agentType
                    }
                    val resolvedGatewayAvailable = event.data.gatewayAvailable ?: current.gatewayAvailable
                    val resolvedGatewayConnected = event.data.gatewayConnected ?: current.gatewayConnected
                    val resolvedGatewayHasError = if (resolvedGatewayAvailable == false && resolvedGatewayConnected != true) {
                        false
                    } else {
                        event.data.gatewayHasError ?: current.gatewayHasError
                    }
                    val clearOpenClawDetails = resolvedGatewayConnected != true
                    stateDebug { "AgentType resolve: event=${event.data.agentType}, current=${current.agentType}, resolved=$resolvedAgentType" }
                    // When the dashboard is pinned to an aggregate primary
                    // (daemon / openclaw) and a sibling (claude / codex / …)
                    // emits a state_update, that sibling's state is *not* the
                    // aggregate's state — it lives in `siblingSessions`. Apply
                    // the incoming state only when we are NOT keeping the
                    // aggregate identity, otherwise the OpenClaw crayfish
                    // animates as if Claude's PROCESSING were its own.
                    val resolvedAgentState =
                        if (keepAggregateIdentity) current.agentState else event.data.state
                    // sessionId on state_update tracks the latest hook-active session.
                    // In aggregate mode (daemon/openclaw), keep the original bridge-core UUID
                    // as sessionId, not a sibling session's ID. This prevents the sibling from
                    // being skipped as "self" when processing siblingSessions for creature rendering.
                    // Also skip sessionId update when the event itself is from an aggregate primary.
                    val isAggregateEvent = event.data.agentType in AGGREGATE_AGENT_TYPES
                    current.copy(
                        agentState = resolvedAgentState,
                        permissionMode = event.data.permissionMode,
                        agentType = resolvedAgentType,
                        currentTool = event.data.currentTool,
                        toolInput = event.data.toolInput,
                        toolProgress = event.data.toolProgress,
                        projectName = if (keepAggregateIdentity) current.projectName else event.data.projectName ?: current.projectName,
                        modelName = if (keepAggregateIdentity) current.modelName else event.data.modelName ?: current.modelName,
                        effortLevel = if (keepAggregateIdentity) current.effortLevel else event.data.effortLevel ?: current.effortLevel,
                        billingType = event.data.billingType ?: current.billingType,
                        options = event.data.options ?: emptyList(),
                        promptType = event.data.promptType,
                        question = event.data.question,
                        suggestedPrompt = event.data.suggestedPrompt,
                        remoteUrl = event.data.remoteUrl ?: current.remoteUrl,
                        agentCapabilities = event.data.agentCapabilities
                            ?: current.agentCapabilities.takeUnless { clearOpenClawDetails && it?.type == "openclaw" },
                        modelCatalog = event.data.modelCatalog
                            ?: current.modelCatalog.takeUnless { clearOpenClawDetails && current.agentType == "openclaw" },
                        sessionStatus = event.data.sessionStatus
                            ?: current.sessionStatus.takeUnless { clearOpenClawDetails },
                        pairingUrl = event.data.pairingUrl ?: current.pairingUrl,
                        navigable = event.data.navigable,
                        cursorIndex = event.data.cursorIndex,
                        workerSessionCount = when {
                            resolvedGatewayConnected != true -> null
                            event.data.workerSessionCount != null -> event.data.workerSessionCount
                            else -> current.workerSessionCount
                        },
                        ollamaStatus = event.data.ollamaStatus ?: current.ollamaStatus,
                        mlxModels = event.data.mlxModels ?: current.mlxModels,
                        subscriptions = event.data.subscriptions ?: current.subscriptions,
                        antigravityStatus = event.data.antigravityStatus ?: current.antigravityStatus,
                        gatewayAvailable = resolvedGatewayAvailable,
                        gatewayConnected = resolvedGatewayConnected,
                        gatewayHasError = resolvedGatewayHasError,
                        moduleHealth = event.data.moduleHealth ?: current.moduleHealth,
                        voiceAssistantState = event.data.voiceAssistantState ?: current.voiceAssistantState,
                        voiceAssistantText = event.data.voiceAssistantText,
                        voiceAssistantResponseText = event.data.voiceAssistantResponseText,
                        // sessionId on state_update tracks the latest hook-active session ("may move with hook activity"
                        // per shared/src/protocol.ts). Mirrors Apple `if let sid = e.sessionId { state.sessionId = sid }`.
                        // Absent → keep current (set originally by BridgeEvent.Connected).
                        // In aggregate mode (daemon/openclaw), keep the original bridge-core UUID as sessionId,
                        // not a sibling session's ID. This prevents the sibling from being skipped as "self"
                        // when processing siblingSessions for creature rendering.
                        sessionId = when {
                            isAggregateEvent || keepAggregateIdentity -> current.sessionId
                            else -> event.data.sessionId ?: current.sessionId
                        },
                        // focusedSessionId carries explicit user-focus state.
                        // Mirror Apple AgentStateHolder.handleStateUpdate:
                        //   absent → keep current (session-bridge updates
                        //     don't carry focus context).
                        //   present empty ("") → daemon explicitly cleared
                        //     focus; propagate as null so highlights drop.
                        //   present non-empty → adopt as the new focus.
                        focusedSessionId = if (event.data.focusedSessionId != null) {
                            event.data.focusedSessionId.takeIf { it.isNotEmpty() }
                        } else {
                            current.focusedSessionId
                        },
                    ).also {
                        stateDebug { "State updated: sessionId=${it.sessionId}, agentType=${it.agentType}, agentState=${it.agentState}, siblingSessions.size=${it.siblingSessions.size}" }
                    }
                }
                lastKnownState = _state.value
                // Apple AgentStateHolder parity: the OpenClaw Gateway provides
                // its own rich timeline entries via timeline_event — suppress
                // the StateTimelineGenerator fallback ("Connected"/"Prompt
                // sent"/tool rows) as soon as the gateway is confirmed
                // connected, so the generator can't race ahead of the first
                // timeline push after (re)connect.
                if (event.data.gatewayConnected == true) {
                    StateTimelineGenerator.instance.setReceivingBridgeTimeline(true)
                }
                StateTimelineGenerator.instance.onStateUpdate(event.data)
                SessionMetrics.instance.onMessageReceived()
            }

            is BridgeEvent.UserPrompt -> {
                StateTimelineGenerator.instance.setLastUserPrompt(event.text)
            }

            is BridgeEvent.Usage -> {
                // Scrub upstream subscription-quota numbers when the daemon
                // flags the data as stale (no live CLI to fetch a fresh
                // value). Every surface that reads state.usage.fiveHourPercent
                // / sevenDayPercent / extraUsage* then collapses on null,
                // matching macOS/Pixoo/D200H/plugin behavior without each
                // downstream view needing its own stale check.
                val incoming = if (event.data.usageStale == true) {
                    event.data.copy(
                        fiveHourPercent = null,
                        sevenDayPercent = null,
                        fiveHourResetsAt = null,
                        sevenDayResetsAt = null,
                        extraUsageEnabled = null,
                        extraUsageMonthlyLimit = null,
                        extraUsageUsedCredits = null,
                        extraUsageUtilization = null,
                    )
                } else event.data
                _state.update { current ->
                    current.copy(
                        usage = incoming,
                        oauthConnected = incoming.oauthConnected ?: current.oauthConnected,
                        ollamaStatus = incoming.ollamaStatus ?: current.ollamaStatus,
                        modelCatalog = incoming.modelCatalog
                            ?: current.modelCatalog.takeUnless {
                                current.gatewayConnected != true && current.agentType == "openclaw"
                            },
                        mlxModels = incoming.mlxModels ?: current.mlxModels,
                        subscriptions = incoming.subscriptions ?: current.subscriptions,
                        antigravityStatus = incoming.antigravityStatus ?: current.antigravityStatus,
                        // codexRateLimits only rides on usage_update — hoist it
                        // so a later null incoming doesn't wipe it (mirrors iOS).
                        codexRateLimits = incoming.codexRateLimits ?: current.codexRateLimits,
                    )
                }
                lastKnownState = _state.value
                SessionMetrics.instance.onMessageReceived()
            }

            is BridgeEvent.Voice -> {
                _state.update { it.copy(voice = event.data) }
            }

            is BridgeEvent.Connected -> {
                // Clear stale timeline (logs) on (re)connect. A fresh daemon with
                // an empty store sends NO `timeline_history`, so the replace path
                // (BridgeEvent.TimelineHistory) never fires and old rows linger.
                // `connection` precedes `timeline_history` in the initial burst,
                // so a daemon that HAS history repopulates it immediately after
                // this clear; a fresh one correctly stays empty.
                TimelineStore.instance.clear()
                _state.update {
                    it.copy(
                        bridgeConnected = true,
                        sessionId = event.sessionId,
                        // Drop any focus carried over from a prior connection.
                        // The daemon's relay re-emits focus on its first
                        // state_update, so re-establishing connection should
                        // start clean.
                        focusedSessionId = null,
                    )
                }
                SessionMetrics.instance.onConnected()
            }

            is BridgeEvent.DisplaySleep -> {
                _state.update { it.copy(hostDisplayOn = event.displayOn, hostDim = event.dim) }
            }

            is BridgeEvent.SessionsList -> {
                stateDebug { "SessionsList: ${event.sessions.size} sessions" }
                if (VERBOSE_STATE_LOGS || Log.isLoggable(TAG, Log.DEBUG)) {
                    for (s in event.sessions) {
                        Log.d(TAG, "  Session: id=${s.id}, agentType=${s.agentType}, state=${s.state}, projectName=${s.projectName}")
                    }
                }
                _state.update { it.copy(siblingSessions = event.sessions) }
            }

            is BridgeEvent.EncoderState -> { /* Deck tab removed — ignore */ }

            is BridgeEvent.ButtonState -> { /* Deck tab removed — ignore */ }

            is BridgeEvent.SlotMap -> { /* Deck tab removed — ignore */ }

            is BridgeEvent.Timeline -> {
                if (event.upsert) {
                    TimelineStore.instance.upsertEntry(event.entry.toTimelineEntry())
                } else {
                    TimelineStore.instance.addEntry(event.entry.toTimelineEntry())
                }
                StateTimelineGenerator.instance.setReceivingBridgeTimeline(true)
            }

            is BridgeEvent.TimelineHistory -> {
                // The daemon's only history push to a client is the full store
                // snapshot on (re)connect — replace, don't merge, so re-stamped
                // OpenClaw rows can't stack up across the tablet's reconnects.
                TimelineStore.instance.replaceSnapshot(event.entries.map { it.toTimelineEntry() })
                StateTimelineGenerator.instance.setReceivingBridgeTimeline(true)
            }

            is BridgeEvent.Disconnected -> {
                _state.update {
                    it.copy(
                        bridgeConnected = false,
                        agentState = AgentState.DISCONNECTED,
                        hostDisplayOn = true,
                        gatewayConnected = false,
                        gatewayHasError = false,
                        workerSessionCount = null,
                        siblingSessions = emptyList(),
                        // Both session ids are connection-scoped — clear so
                        // a fresh connection starts clean (mirrors Apple
                        // AgentStateHolder.resetToDisconnected).
                        sessionId = null,
                        focusedSessionId = null,
                        // Purge Node-relayed usage/LIMITS cache. A fresh (or
                        // limited) daemon on reconnect never re-emits these, and
                        // the top-level fields use null-coalescing merge that
                        // would otherwise keep the OLD daemon's values frozen —
                        // showing LIMITS/subscription data that is actually
                        // unavailable. Cleared on the DISCONNECT edge (not
                        // Connect) because the reconnect burst delivers
                        // usage_update BEFORE the connection event, so the fresh
                        // daemon's real values repopulate cleanly right after.
                        // Mirrors Apple AgentStateHolder.clearRelayedUsageState().
                        usage = UsageUpdate(),
                        subscriptions = emptyList(),
                        antigravityStatus = null,
                        codexRateLimits = null,
                    )
                }
                SessionMetrics.instance.onDisconnected()
                StateTimelineGenerator.instance.onDisconnected()
            }
        }
    }
}
